### ISSUE TITLE
Ghidra disassembler written in pure Golang. Allow running on Linux bare metal without a VM. 

### ISSUE THREAD

**Issue by Sasson22** [NationalSecurityAgency/ghidra#371]
GoLang source code is requested to harness Ghidras powerful RCE framework to scale for concurrency applications and multi threading. Ghidra's future adaptation in RCE should address eminent issues for a robust life cycle.

Specifically, concurrence and mutli threading is rapidly emerging and not currently served by the SAXParser, recently adapted after 9.0.

SAXParser class is not thread-safe. Therefore, multiple threads must not use the same SAXParser instance at the same time. ... Each thread has one SAXParser instance. NodeJS is single threaded. Go is multi threaded. Java and JVM have inherent security vulnerabilities and fail to scale.

Golang is thread safe, memory safe and writes to machine code on Linux without a VM. Golang is requested to further address efficiency and robust downstream adaptation, and provide secure remote debugging in a multi threaded reality.

-Golang is faster than Node. Further this request facilitates in analyzing and integrating propriety code, IoT server-client relations and data analytics.

-Specific, initial feature request should run a Linux Kernel x86_64 (since Raspberry Pi can currently run this). This further allows for conversion into debian packages, consequently followed by Ubuntu, etc. for greater distribution and crowd sourcing development.

Harnessing this valuable public offering is strengthened.

**Comment by saruman9** [comment]
Please explain why not Rust or Ada or C++17/20?

**Comment by vmcall** [comment]
lol no generics

**Comment by Sasson22** [comment]
> Please explain why not Rust or Ada or C++17/20?
Good inquiry, rather succinctly:

C++ ruled out, as production debug builds are not seen as optimal relevant to security compliance.
Go first requested due to downstream crowd development adaptation, designed to enable more efficient communication between programmers and a robust “done right”  approach championed worldwide, adding to Ghidras valuable public offering usage growth. Golang user request features are built at an impressive pace including Generics or parametric polymorphism features being promised already and coming in with Go2. Functional Programming advances over OOL.

Rust is impressive, while not considered an optimal server side option and ranks lower in many public surveys and usage charts. In what appears to be public open source development disinterest. It is seen as superior in memory management and also more complex in smooth adaptation to diverse IoT interconnections, perhaps due to strict compile-time data ownership mechanisms.

Reconsideration with proper metrics and new information is keeping Rust relevant. And further developments such as Haskell 2020 will require reconsideration as time goes by, with lazy evaluations holding advances in a purists dream.

In summary, Goland and soon to be release GO2 (with Generics) is seen as optimal for Ghidra to scale and achieve greater acceptance and use in functional, supported and actually used by the public, open source world.

Finally, Ghidra loaning greater debugging possibilities that may address, performance issue of debugging optimized binaries, secure remote debugging while remaining memory safe, multi thread safe, scalable with concurrency. All while serving rich development exposure and public use in the current reality.

**Comment by Sasson22** [comment]
> lol no generics

Go2 promising generics...

**Comment by JustasMasiulis** [comment]
Really not sure if intentional troll or one of those delusional `rewrite x in y` shills.

> Rust is impressive, while not considered an optimal server side option

Considered by whom?

> and ranks lower in many public surveys and usage charts.

Ranks lower by what measure. Show the numbers.

> In what appears to be public open source development disinterest.

I don't see any disinterest of open source development in rust. Plenty of projects and libraries exist.

> In summary, Goland and soon to be release GO2 (with Generics) is seen as optimal for Ghidra to scale and achieve greater acceptance and use in functional, supported and actually used by the public, open source world.

Seen as optimal by whom?

You also forgot to _reconsider_ modern c++ in the current _reality_ of your post.

**Comment by Sasson22** [comment]
**Value adding comments appreciated.  

Clarity of purpose. Value recognition of Ghidra framework and toolkit for greater public recognition. If you have a value adding perspective share it. Your opinion is appreciated and functionally considered without hierarchies obfuscating designs, suggestions or requests.  If you feel entitled to an authority position perhaps you deserve it. Can you keep the discourse and free flow of sharing functional?

This issue is about building in enhancement. Ghidra is perhaps the greatest open source public offering of our lifetime, imo. It is growing quickly, despite the XXE issue in XML parsing in initial release. And it continues to expand, by example, today it is operational on Docker. Further, keeping the ball rolling...

Being in the public domain with source code freely available and viable to broad spectrum crowd-sourcing. Further, this smart move is bringing more people up to speed on this powerful framework and the impressive tool-kits! The younger generation is interested imo towards results that are functional, able to build career credits, and document contributions that are initially given freely. 

Fostering such a culture here to allow greater cooperation among interested individuals, further adds value to the project imo.

And in that vein continuing accordingly...**

> > Rust is impressive, while not considered an optimal server side option
> 
> Considered by whom? 

**imo...look at the stacks built for such purposes if you like.**
> 
> > and ranks lower in many public surveys and usage charts.

> 
> Ranks lower by what measure. Show the numbers.

**Here's a link for you...https://stackshare.io/stackups/go-vs-haskell-vs-rust**

> 
> > In what appears to be public open source development disinterest.
> 
> I don't see any disinterest of open source development in rust. Plenty of projects and libraries exist.

**Thank you for your opinion. Velocity metrics may help here. Again imo. Granted, no disrespect to Rust. Rust is impressive and appears to be daunting for many. And by all means, please integrate your value adding enhancements at will. I still feel there is a general sense of intimidation out there in relation to Rust. Immediate results are not as rewarding and technical hang ups abound imo. Hence the interest factor is not seen as interesting.** 
> 
> > In summary, Goland and soon to be release GO2 (with Generics) is seen as optimal for Ghidra to scale and achieve greater acceptance and use in functional, supported and actually used by the public, open source world.
> 
> Seen as optimal by whom?

**Public use and experimental development by a larger cross section of people capable of adding value. As noted for example: user requests -are listened to -in Go developments. Generics are being integrated in ways that are simple to implement and be as functional as possible. Have you seen such traits in Rust or Haskell? I respect both greatly and am a fan. Perhaps you can enlighten me?**

> 
> You also forgot to _reconsider_ modern c++ in the current _reality_ of your post.

> Ah... You are fast. I edited this into my response very shortly after first hitting the comment button.. You miss this..."C++ 17/21 ruled out, as production debug builds are not seen as optimal relevant to security compliance."  And if you are so quick to respond, I would definitely want you on my team! Albeit I probably can afford you! Cheers.

**Comment by vmcall** [comment]
have you considered rewriting the linux kernel in go, and directly implementing ghidra, to scale for concurrency applications and multi threading? can we add this to the suggestion list? this would make ghidra super fast, no generics and no context switches!!

**Comment by brunocassol** [comment]
> Really not sure if intentional troll

Regardless, it's going to be full of trolls now as someone submitted this to a troll Reddit sub:
https://old.reddit.com/r/programmingcirclejerk/comments/bamxxd/ghidra_disassembler_written_in_pure_golang_allow/

I'd suggest restricting this conversation to NSA members and Issue author only.

**Comment by vmcall** [comment]
restriction of debate is a breach of free speech! this is a public, open-source repository where everyone ( go-developers and mentally stable) should be allowed to criticize not only code but also the issues brought up.

**Comment by brunocassol** [comment]
> have you considered rewriting the linux kernel in go
> restriction of debate is a breach of free speech

Yeah, as I said. This is going to be full of trolls now. Thanks for proving my point @vmcall

**Comment by vmcall** [comment]
imagine having jabba repos...webscale was a mistake

**Comment by jha** [comment]
I don't think Golang is low level enough, this should be done in Verilog. Imagine the benefits that a super parallelized application running on an FPGA could offer over slower general-purpose processors. In fact the NSA team should begun to develop ASICs exclusively for the purpose of analysis.
